kernel: Create usable static kernel library

sedited opened the pull request:
Since the move to cmake, the kernel static library that is installed after a cmake --install build is unusable. It lacks symbols for the internal libraries, besides those defined in the kernel library target.

Fix this by explicitly installing all the required internal static libraries. To make usage of these installed libraries easy, add a pkg-config file that can be used during linking.

This patch can be tested with:

```
cmake -B build -DBUILD_SHARED_LIBS=OFF -DBUILD_KERNEL_LIB=ON
cmake --build build
cmake --install build
g++ -std=c++20 -o test_chainstate src/bitcoin-chainstate.cpp -I/home/drgrid/bitcoin/src $(pkg-config --libs --static libbitcoinkernel)
```

Attempts to solve #30801

ryanofsky reviewed (approved):
Code review ACK 192d71e312d0eab6cc1ce061edb34d9d338f7fec.

This seems like a reasonable approach, but I wonder if a more consistent way of doing this would just treat all the libraries the same as secp, installing them to the lib/ directory and listing them in libbitcoinkernel.pc. That way the secp library would not be an outlier, the TARGET_OBJECTS calls would not be needed, and build would be lighter, with each .o object file added to one library file .a instead of two.

TheCharlatan commented:
Not sure yet why the CI fails, will investigate and leave this in draft for now. If it is something cmake version related, I'll probably end up abandoning this approach and instead try to get the installation of all the required libraries working.

sedited committed:
build: Produce a usable static kernel library

Since the move to cmake, the kernel static library that is installed
after a cmake --install build is unusable. It lacks symbols for the
internal libraries, besides those defined in the kernel library target.

This is because cmake, unlike the libtool archiver, does not combine
multiple static libraries into a single static library on installation.
This is likely an intentional design choice, since there were a bunch of
caveats to the way libtool calculated these libraries.

Fix this problem by installing all the required libraries. The user must
then link all of them along with the bitcoin kernel library.

sedited committed:
build: Add a pkg-config file for libbitcoinkernel

TheCharlatan commented:
Re-worked this, it now installs all the required libraries when the static kernel lib is being built. Couldn't figure out why the previous method was working on my machines, but not in the CI container. Since it is not a standard thing to do (just about touching undefined behaviour territory), and some reviewers commented they would support installing all libraries as an alternative, this now seems like the better and future-proof solution to me.

ryanofsky commented on src/kernel/CMakeLists.txt:
```diff
@@ -98,6 +98,32 @@ set_target_properties(bitcoinkernel PROPERTIES
+  get_target_static_link_libs(bitcoinkernel all_target_static_link_libs)
+
+  foreach(lib ${all_target_static_link_libs})
+    install(TARGETS ${lib} ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
```
In commit "build: Produce a usable static kernel library" (7739fa69462f070c62345f2f14fcf817a512d59a)

Seems ilke in theory this could install the same libraries more than once because it is just recursively adding dependencies without deduplicating them. For example if kernel library depends on util and crypto, and util library depends on crypto, crypto library could be listed twice. May be worth iterating over a unique version of the list so the install step is cleaner if it does contain duplicates

TheCharlatan replied:
I was doing this, but removed it again from the patch. I feel like if there are duplicates, it is a bug in our library topology. But it could still make sense to guard against them.

ryanofsky replied:
Yeah I would leave alone for now to keep things simple. I don't think it is a bug to have multiple libraries a target depends on depend on some of the same libraries themselves though.

ryanofsky reviewed (approved):
Code review ACK 766881e33d13ca2887f7ed179cdcc6bc007a6325. I like this approach, seems more generic and lightweight.

hebasto commented:
766881e33d13ca2887f7ed179cdcc6bc007a6325

While installing `libbitcoinkernel.pc`, why ignore CMake's configs for downstream projects using CMake?

hebasto commented on src/kernel/CMakeLists.txt:
```diff
@@ -98,6 +98,39 @@ set_target_properties(bitcoinkernel PROPERTIES
 
+# When building the static library, install all static libraries the
+# bitcoinkernel depends on.
+if (NOT BUILD_SHARED_LIBS)
```
style nit: All other CMake code uses the `if()` command without an additional space.

hebasto reviewed (commented):
Approach ACK 766881e33d13ca2887f7ed179cdcc6bc007a6325.

hebasto commented on src/kernel/CMakeLists.txt:
```diff
@@ -98,6 +98,39 @@ set_target_properties(bitcoinkernel PROPERTIES
+  set(all_target_static_link_libs)
+  get_target_static_link_libs(bitcoinkernel all_target_static_link_libs)
+
+  set(LIBS_PRIVATE)
```
```suggestion
  set(LIBS_PRIVATE "")
```
if there is no intention to let the user set it as a cache variable.

(same for `all_target_static_link_libs` a few lines above)

Why this local variable name capitalized?

TheCharlatan replied:
It is capitalized to match the substitution in the pkg-config file. I had the impression the substitution variables are usually capitalized as a convention.

hebasto commented on src/kernel/CMakeLists.txt:
```diff
@@ -98,6 +98,39 @@ set_target_properties(bitcoinkernel PROPERTIES
+  set(LIBS_PRIVATE)
+  foreach(lib ${all_target_static_link_libs})
+    install(TARGETS ${lib} ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+    set(LIBS_PRIVATE "${LIBS_PRIVATE} -l${lib}")
```
style: A shorter alternative:
```suggestion
    string(APPEND LIBS_PRIVATE " -l${lib}")
```

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/30814#issuecomment-2334333003Z

> While installing libbitcoinkernel.pc, why ignore CMake's configs for downstream projects using CMake?

Do you mean why I am not contributing a cmake configuration file as well in this pull request? Or is there something wrong with the current installation step that precludes also installing a configuration file? I just did not have the need for a cmake configuration file so far,. It would be good to contribute one eventually as well though. I could do it now as well, if you think it should be part of this pull request.

TheCharlatan commented:
Updated 766881e33d13ca2887f7ed179cdcc6bc007a6325 -> 0dd16d7118f10ac291a45644769121cbdfd2352f ([staticKernel_0](https://github.com/TheCharlatan/bitcoin/tree/staticKernel_0) -> [staticKernel_1](https://github.com/TheCharlatan/bitcoin/tree/staticKernel_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/staticKernel_0..staticKernel_1))

* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/30814#discussion_r1747381222), fixed spacing around `if` statements.
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/30814#discussion_r1747393816), set variables with `""`.
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/30814#discussion_r1747397241), use `string(APPEND...` to append to the link string.

hebasto reviewed (approved):
ACK 0dd16d7118f10ac291a45644769121cbdfd2352f.

fanquake reviewed (approved):
ACK 0dd16d7118f10ac291a45644769121cbdfd2352f - this looks like a good place to start.

ryanofsky reviewed (approved):
Code review ACK 0dd16d7118f10ac291a45644769121cbdfd2352f

Just cmake style improvements since last reveiw

hebasto commented on src/kernel/CMakeLists.txt:
```diff
@@ -98,6 +98,40 @@ set_target_properties(bitcoinkernel PROPERTIES
+endif()
+
+configure_file(${PROJECT_SOURCE_DIR}/libbitcoinkernel.pc.in ${PROJECT_BINARY_DIR}/libbitcoinkernel.pc @ONLY)
+install(FILES ${PROJECT_BINARY_DIR}/libbitcoinkernel.pc DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
```
Should this be a part of the ["Kernel" component](https://github.com/bitcoin/bitcoin/pull/30835)?

TheCharlatan replied:
Yeah, I missed this. Already patched in #30595 , but could also patch it independently if you want to.

hebasto replied:
> Already patched in #30595 , but could also patch it independently if you want to.

Great!

Feel free to consider another refactor commit to get rid off `foreach()` loop: https://github.com/hebasto/bitcoin/commits/240915-kernel/.
